### PR TITLE
More trivial cost matrix fixes....

### LINKED ISSUE #4998: Changing the Order of Targets in the Matrix Causes Incorrect Time and Distance Calculations
We have identified a significant issue with the matrix computation algorithm. It appears that the order of targets in the matrix request is causing incorrect computations. This behavior might be linked to how trivial routes are handled, as the incorrect distances in the matrix seem to be associated with trivial routes.

We have dedicated considerable time to isolate this issue and provide the smallest possible example to demonstrate the erroneous behavior.

Matrix A=>B   result 4 seconds 33 meters correct result

request:

```json
{
	"sources": [
		{
			"lat": 49.39571,
			"lon": 15.57653
		}
	],
	"targets": [
		{
			"lat": 49.39582,
			"lon": 15.57611
		}
	],
	"verbose": false,
	"costing": "auto"	
}
```
result:
```json
{
    "algorithm": "costmatrix",
    "units": "kilometers",
    "sources_to_targets": {
        "durations": [
            [
                4
            ]
        ],
        "distances": [
            [
                0.033
            ]
        ]
    }
}
```

Matrix A=>C   result 100 seconds 909 meters correct result

request:

```json
{
	"sources": [
		{
			"lat": 49.39571,
			"lon": 15.57653
		}
	],
	"targets": [
		{
			"lat": 49.39564,
			"lon": 15.57676
		}
	],
	"verbose": false,
	"costing": "auto"	
}
```

result:

```json
{
    "algorithm": "costmatrix",
    "units": "kilometers",
    "sources_to_targets": {
        "durations": [
            [
                100
            ]
        ],
        "distances": [
            [
                0.909
            ]
        ]
    }
}
```
Matrix A=> {B,C} result {4 seconds 33 meters, 100 second 909 meters} everything is perfect same as simple matrices => correct results

request:
```json
{
	"sources": [
		{
			"lat": 49.39571,
			"lon": 15.57653
		}
	],
	"targets": [
        {
			"lat": 49.39582,
			"lon": 15.57611
		},
        {
			"lat": 49.39564,
			"lon": 15.57676
		}
        
	],
	"verbose": false,
	"costing": "auto"
}
```
result:
```json
{
    "algorithm": "costmatrix",
    "units": "kilometers",
    "sources_to_targets": {
        "durations": [
            [
                4,
                100
            ]
        ],
        "distances": [
            [
                0.033,
                0.909
            ]
        ]
    }
}
```

Matrix A=> {C,B} result {100 seconds 909 meters, 106 second 960 meters}  A=>C element is correct A=>B element is completely nonsense instead of 4 seconds we have 106 seconds and instead od 33 meters we have 960 meters

request:
```json
{
	"sources": [
		{
			"lat": 49.39571,
			"lon": 15.57653
		}
	],
	"targets": [
        {
			"lat": 49.39564,
			"lon": 15.57676
		},
        {
			"lat": 49.39582,
			"lon": 15.57611
		}
        
	],
	"verbose": false,
	"costing": "auto"
}
```
result:
```json
{
    "algorithm": "costmatrix",
    "units": "kilometers",
    "sources_to_targets": {
        "durations": [
            [
                100,
                106
            ]
        ],
        "distances": [
            [
                0.909,
                0.960
            ]
        ]
    }
}
```

### PR DESCRIPTION
**Pull request by chrstnbwnkl**
# Issue

Fixes #4998 

This time I can't blame @nilsnolde :sweat_smile: this one's on me. CostMatrix was returning faulty routes in trivial cases where: 
  1. at least one source and at least two target locations are correlated to the same edge 
  2. at least one of the targets is located further down, and at least one target further up along that edge than a source location 

This PR fixes this edge case. And while at it, I made some time dependent matrix tests I wrote earlier this year a bit more concise. 

## Tasklist

 - [x] Add tests
 - [x] Add #fixes with the issue number that this PR addresses
 - [x] Update the [changelog](CHANGELOG.md)

### CODE REVIEW

File: src/thor/costmatrix.cc [hunk 1]
```diff
@@ -923,7 +923,8 @@ void CostMatrix::CheckReverseConnections(const uint32_t target,
       // Special case - common edge for source and target are both initial edges
       if (rev_pred.predecessor() == kInvalidLabel && fwd_predidx == kInvalidLabel) {
         // bail if the edge wasn't allowed
-        if (!rev_pred.path_id()) {
+        if (!fwd_label.path_id()) {
```
Comment (chrstnbwnkl) [inline src/thor/costmatrix.cc:926]: On the reverse edge label, we store the allowed flag in `path_id` for the _opposing edge_, we should be looking at the forward label's `path_id` here.

File: test/gurka/test_matrix.cc [hunk 2]
```diff
@@ -41,7 +41,7 @@ void check_matrix(const rapidjson::Document& result,
                         std::to_string(i % origin_td.Size());
       EXPECT_NEAR(v.GetObject()["distance"].GetFloat(), exp_dists[i], 0.01) << msg;
       if (valid_traffic) {
-        EXPECT_TRUE(v.GetObject().HasMember("date_time")) << msg;
+        ASSERT_TRUE(v.GetObject().HasMember("date_time")) << msg;
```
Comment (chrstnbwnkl) [inline test/gurka/test_matrix.cc:44]: some small improvements, the tests were failing locally, so I made them a bit more comprehensible.

File: test/gurka/test_matrix.cc [hunk 3]
```diff
@@ -92,7 +92,7 @@ class MatrixTrafficTest : public ::testing::Test {
                             {{"service_limits.max_timedep_distance_matrix", "50000"},
                              {"mjolnir.traffic_extract",
                               "test/data/matrix_traffic_allowed/traffic.tar"},
-                             {"mjolnir.timezone", "test/data/tz.sqlite"},
+                             {"mjolnir.timezone", VALHALLA_BUILD_DIR "test/data/tz.sqlite"},
```
Comment (chrstnbwnkl) [inline test/gurka/test_matrix.cc:95]: this was the reason they were failing on my machine

File: test/gurka/test_matrix.cc [hunk 4]
```diff
@@ -916,3 +861,75 @@ TEST(StandAlone, HGVNoAccessPenalty) {
     }
   }
 }
+
+/************************************************************************ */
```
Comment (chrstnbwnkl) [inline test/gurka/test_matrix.cc:865]: Here we get to the relevant stuff: I added a test case that mimics the situation documented excellently by @radekvermirovsky in #4998 and make sure it now passes, with and without reverse connection checks.

File: test/gurka/test_matrix.cc (hunk 3 above)
Comment (kevinkreiser) [reply test/gurka/test_matrix.cc:95]: yeah i had a branch (long stale) that fixed all of these but there were other thigns on it nils suggested that i never got around to

File: src/thor/costmatrix.cc [hunk 5]
```diff
@@ -1164,7 +1164,7 @@ void CostMatrix::SetTargets(baldr::GraphReader& graphreader,
                              !opp_dir_edge->not_thru(), !(costing_->IsClosed(directededge, tile)),
                              static_cast<bool>(flow_sources & kDefaultFlowMask),
                              InternalTurn::kNoTurn, kInvalidRestriction,
-                             static_cast<uint8_t>(costing_->Allowed(opp_dir_edge, opp_tile)),
+                             static_cast<uint8_t>(costing_->Allowed(directededge, tile)),
```
Comment (kevinkreiser) [inline src/thor/costmatrix.cc:1167]: which field of the edgelabel is this, hard to review this part

Comment (chrstnbwnkl) [reply src/thor/costmatrix.cc:1167]: this is `path_id`, the one that's used to store whether the edge is allowed in the initial edge labels

**Review by kevinkreiser — APPROVED** [review]
